### PR TITLE
VariantContext to BQ row(Implementation and Test)

### PR DESCRIPTION
**Pull request by BoscoSuen**
This PR provides one feature: generating BQ row from VariantContext. It includes many codes(implementation, unit tests and integration tests) and it is pretty hard to separate utility functions and BQ row generator into 2 PRs which will make design logic hard to understand, please find my implementation and test logic in the following summary:
## Implementation
- **Pipeline Runner**: creating a Service interface **BigQueryRowGenerator.java**(and also its implementation class) and use the getRows() method in the interface to generate a **PCollection\<TableRow\>** in the pipelineRunner. The header information comes from VcfToBqContext.
- **Generating BigQuery Row from VaraintContext(single VCF record)**: Creating a TableRow as a base row, and set all the field values from VariantContext, the tableRow should contain some nested rows for ALT fields and Calls(genotypes). More detailed notes are provided in my method comments. 
- **VariantToBqUtils**: Note that in the VariantContext, the values are decoded as string, thus we need to convert them to right value type defined by VCF header. Also, some empty fields(".") should be converted to null in the BQ row. This Utility class includes all utility function for storing constant field names, methods to convert value type and deal with some field splitting works(*some fields in INFO field with number=A should be moved to ALT field*)
- **Note**: sanitizing fields and BigQueryRow generator flags are not implemented, they can be added later if needed.

## Testing plan
- **Unit test**: mock all the field values in VariantContext(alleles, genotypes), and test all the methods in the **VariantToBqUtils.java** including the utility function for converting value type, splitting info field into ALT field, constructing Genotypes methods is correct.
- **Integration test**: Read a sample VCF file, inject the **vcfParser**, **bigQueryRowGenerator**, test the tranformation from VCF file to VCFHeader, VCF record -> VariantContext -> Bigquery row. Test the row elements. Some corner cases testing are included.
- **Usage**: ```./gradlew test```

### CODE REVIEW

File: src/main/java/com/google/gcp_variant_transforms/beam/helper/ConvertVariantToRowFn.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/** {@link DoFn} implementation for a VariantContext to a BigQuery Row. */
+public class ConvertVariantToRowFn extends DoFn<VariantContext, TableRow> {
+  private final boolean allowIncompatibleRecords = false;
+  private final boolean omitEmptySampleCalls = false;
+  private BigQueryRowGenerator bigQueryRowGenerator;
+  private VCFHeader vcfHeader;
+
+  public ConvertVariantToRowFn(VCFHeader vcfHeader, BigQueryRowGenerator bigQueryRowGenerator) {
+    this.bigQueryRowGenerator = bigQueryRowGenerator;
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/beam/helper/ConvertVariantToRowFn.java:19]: Should we set the private variables in the same order as they're listed as parameters, i.e. vcfHeader then bigQueryRowGenerator

File: src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorImpl.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    // write alt field and info field to BQ row
+    List<TableRow> altMetadata = VariantToBqUtils.addAlternates(variantContext, alleleIndexingMap);
+    // altMeta should add those field that info number = 'A', then add to row
+    // Other fields will directly add to the base BQ row
+    VariantToBqUtils.addInfo(row, variantContext, altMetadata, vcfHeader);
+    row.set(VariantToBqUtils.ColumnKeyConstants.ALTERNATE_BASES, altMetadata);
+
+    row.set(VariantToBqUtils.ColumnKeyConstants.QUALITY, variantContext.getPhredScaledQual());
+
+    Set<String> filters = variantContext.getFilters();
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorImpl.java:46]: This function is really large, so maybe we could break it up with some helper functions. I'm thinking the creation of the alleleIndexingMap and filters HashSet could be good pieces to move into separate functions. It might help with readability, unless you think this will overcomplicate the code.

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+  private static final String MISSING_FIELD_VALUE = ".";
+  private static final String PHASESET_FORMAT_KEY = "PS";
+  private static final String DEFAULT_PHASESET = "*";
+  private static final int MISSING_GENOTYPE_VALUE = -1;
+
+  /**
+   * Constants for column names in the BigQuery schema.
+   */
+  public static class ColumnKeyConstants {
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:32]: Generating the schema uses these ColumnKeyConstants. Maybe we could rename this class to indicate the utilities here are used in multiple places (not just VariantToBq); alternatively, we could take out just the static class ColumnKeyConstants and put it in a shared utility class for VariantToBq and SchemaGeneration

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    public static final String CALLS = "call";
+    public static final String CALLS_NAME = "name";
+    public static final String CALLS_GENOTYPE = "genotype";
+    public static final String CALLS_PHASESET = "phaseset";
+  }
+
+  public static String buildReferenceBase(VariantContext variantContext) {
+    Allele referenceAllele = variantContext.getReference();
+    // If the ref length is longer than 1, store the ref as a String
+    return dealWithMissingFieldValue(referenceAllele.getDisplayString());
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:51]: Is it necessary to put the ternary operator in a separate function here? The comment confused me at first since we don't see the if statement logic until reading down to dealWithMissingFieldValue()

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  }
+
+  private static void addGenotypes(TableRow row, List<Allele> alleles,
+                                   Map<String, Integer> alleleIndexingMap) {
+    List<Integer> genotypes = new ArrayList<>();
+    for (Allele allele : alleles) {
+      String alleleStr = allele.getDisplayString();
+      if (alleleStr.equals(MISSING_FIELD_VALUE)) {
+        genotypes.add(-1);
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:174]: perhaps we could use a constant here instead of hard coding -1

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    } else {
+      row.set(ColumnKeyConstants.CALLS_PHASESET, DEFAULT_PHASESET);
+    }
+  }
+
+
+  /**
+   * In some fields that values are MISSING_FIELD_VALUE('.'), should set null in BQ row.
+   */
+  private static String dealWithMissingFieldValue(String value) {
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:212]: if we choose to keep this function, what about calling it handleMissingFieldValue(...)?

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+  public static String buildReferenceBase(VariantContext variantContext) {
+    Allele referenceAllele = variantContext.getReference();
+    // If the ref length is longer than 1, store the ref as a String
+    return dealWithMissingFieldValue(referenceAllele.getDisplayString());
+  }
+
+  public static String buildNames(VariantContext variantContext) {
+    // If the ID field is ".", should write null into BQ row
+    return dealWithMissingFieldValue(variantContext.getID());
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:56]: same comment here as the one above for buildReferenceBase(...)

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    } else {
+      // deal with list of values
+      List<Object> valueList = (List<Object>)value;
+      List<Object> convertedList = new ArrayList<>();
+      for (Object val : valueList) {
+        convertedList.add(convertSingleObjectToDefinedType(val, type));
+      }
+      return convertedList;
+    }
+  }
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:138]: add a line break here

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Convert the value to the type defined in the metadata
+   * If the value is an instance of List, iterate the list
+   * and call convertSingleObjectToDefinedType() to convert every value
+   */
+  public static Object convertToDefinedType(Object value, VCFHeaderLineType type) {
+    if (!(value instanceof List)) {
+      // deal with single value
+      // There are some field value is a single String with ',', eg: "HQ" ->"23,27"
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:122]: - maybe reword this comment for clarity, 
e.g. Some field values are a single, comma-separated String (e.g. "HQ" -> "23,27"

- Google Java style question: I'm not sure if multiple line comments should be placed into a block comment like 
/** text 
text 
text **/ 
or left as single line comments // @tneymanov

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+      }
+    }
+    if (phaseSet == null || phaseSet.length() != 0) {
+      // phaseSet is presented(MISSING_FIELD_VALUE('.') or real value)
+      row.set(ColumnKeyConstants.CALLS_PHASESET, phaseSet);
+    } else {
+      row.set(ColumnKeyConstants.CALLS_PHASESET, DEFAULT_PHASESET);
+    }
+  }
+
```
Comment (mjfeigles) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:207]: remove a line here

File: src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  private static final String TEST_FILE = "src/test/resources/vcf_files_small_valid-4.0.vcf";
+  private VCFCodec vcfCodec;
+  private VCFHeader vcfHeader;
+  private ImmutableList<String> vcfRecords;
+  private List<TableRow> rows;
+
+  @Rule
+  public final GuiceBerryRule guiceBerry = new GuiceBerryRule(TestEnv.class);
+
+  @Inject
```
Comment (mjfeigles) [inline src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java:44]: I think we could make these single lines:

@Inject public VcfParser vcfParser;

File: src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    assertThat(firstRow.get(VariantToBqUtils.ColumnKeyConstants.START_POSITION))
+            .isEqualTo(14370);
+    assertThat(firstRow.get(VariantToBqUtils.ColumnKeyConstants.END_POSITION))
+            .isEqualTo(14370);
+    assertThat(firstRow.get(VariantToBqUtils.ColumnKeyConstants.NAMES))
+            .isEqualTo("rs6054257");
+    assertThat(firstRow.get(VariantToBqUtils.ColumnKeyConstants.REFERENCE_BASES)).isEqualTo("G");
+
+    // AF field has type=`A` in VCFHeader, which should be moved from Info field to ALT field
+    List<TableRow> altMetadata = (List<TableRow>) firstRow.get(
```
Comment (mjfeigles) [inline src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java:102]: This test is very large. Is there a way to break up the asserts into multiple functions?
ex. function ideas: 
- test for first row constant fields
- test for first row filter and NS/DP/DB/H2
- test for calls
- test the record with empty fields

File: src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+    // Test the fourth record in the test comment, there are some empty field in the record
+    TableRow fourthRow = rows.get(3);
+    assertThat(fourthRow.get(VariantToBqUtils.ColumnKeyConstants.NAMES)).isEqualTo(null);
+    // ALT field is ".", which should be null in the BQ row
+    List<TableRow> emptyAltMetadata = (List<TableRow>) fourthRow.get(
+            VariantToBqUtils.ColumnKeyConstants.ALTERNATE_BASES);
+    TableRow emptyAltBase = emptyAltMetadata.get(0);
+    assertThat(emptyAltBase.get(VariantToBqUtils.ColumnKeyConstants.ALTERNATE_BASES_ALT))
+            .isEqualTo(null);
```
Comment (mjfeigles) [inline src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java:138]: Are there any other rows we should test in addition to first and fourth row? Maybe some edge cases? 

I am not sure of any, just making sure we consider this.

File: src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+  public void constructMockClasses() {
+    variantContext = mock(VariantContext.class);
+    vcfHeader = mock(VCFHeader.class);
+    genotypesContext = mock(GenotypesContext.class);
+  }
+
+  @Test
+  public void testConvertStringValueToRightValueType_whenComparingElement_thenTrue() {
+    // test list of values
+    String integerListStr = "23,27";
```
Comment (mjfeigles) [inline src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java:54]: I suggest arranging both lists before the asserts. I'm wondering if it would be good to go even further and arrange all the variables in this test function before all the asserts.

File: src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    // test String values
+    String str = "T";
+    assertThat(VariantToBqUtils.convertToDefinedType(str, VCFHeaderLineType.String))
+            .isEqualTo("T");
+  }
+
+  @Test
+  public void testBuildReferenceBase_whenComparingElement_thenTrue() {
+    Allele refAllele = mock(Allele.class);
+    when(refAllele.getDisplayString()).thenReturn("A");   // mock reference value
```
Comment (mjfeigles) [inline src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java:80]: should we save "A" as an expectedString variable to eliminate human error between setting it as the return and checking isEqualTo? Or maybe this is okay since we're not passing in "A" to any function calls being tested--not sure.

File: src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+            alleleIndexingMap);
+    TableRow row = calls.get(0);
+    assertThat(row.get(CALLS_NAME)).isEqualTo("sample");
+    assertThat(row.get(CALLS_GENOTYPE)).isEqualTo(Arrays.asList(1, 2));
+    assertThat(row.get(CALLS_PHASESET)).isEqualTo("0");
+    // PS should not be in the info map, should be set in phase set
+    assertThat(!row.containsKey("PS")).isTrue();
+    assertThat(row.get("HQ")).isEqualTo(Arrays.asList(23,27));
+
+    // Second record: contains empty fields
```
Comment (mjfeigles) [inline src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java:224]: Can we break this test into multiple tests as well?
Maybe a test function for asserting the full table row fields and another one for testing the record containing empty fields?

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java (hunk 3 above)
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:32]: Agree. Have you add a Constants class in your PR? If so I will follow the code you have written and make change in my PR

File: src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorImpl.java (hunk 2 above)
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorImpl.java:46]: Separated adding filter into one util function. Creating alleleIndexingMap only costs one line so I keep the original way

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java (hunk 6 above)
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:212]: I will delete this method, looks useless to separate this into one function

File: src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java (hunk 13 above)
Comment (BoscoSuen) [reply src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java:138]: I have modified the VCF file to put all edge cases into one record row, which could make test more neat

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java (hunk 4 above)
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:51]: Deleted the function

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java (hunk 3 above)
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:32]: I have created a common/Constants.java to store all constants and field names

File: src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java (hunk 12 above)
Comment (BoscoSuen) [reply src/test/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorTest.java:102]: Separated

File: src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java (hunk 14 above)
Comment (BoscoSuen) [reply src/test/java/com/google/gcp_variant_transforms/library/VariantToBqUtilsTest.java:54]: I have moved other field or values to constants in the test, but in this method I think adding custom input value directly in the function will be more clear about the change of the value type. Put these value to static constants naming looks a bit hard to find the value change.

File: src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java [hunk 17]
```diff
@@ -2,46 +2,60 @@
...
+            pipeline.apply(TextIO.read().from(context.getInputFile()))
       .apply(Filter.by((String inputLine) -> !inputLine.startsWith("#")))
-      .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())))
-      .apply(
+      .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())));
+
+    PCollection<TableRow> tableRowPCollection = variantContextPCollection
+            .apply("VariantContextToBQRow",
+                    ParDo.of(new ConvertVariantToRowFn(bigQueryRowGenerator,
+                            context.getVCFHeader())));
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java:52]: add space for tabulation.

File: src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java [hunk 18]
```diff
@@ -2,46 +2,60 @@
...
     this.vcfParser = vcfParser;
+    this.bigQueryRowGenerator = bigQueryRowGenerator;
   }
 
   public void runPipeline() {
     // Demo code.
     Pipeline pipeline = Pipeline.create(options);
-    pipeline.apply(TextIO.read().from(context.getInputFile()))
+    PCollection<VariantContext> variantContextPCollection =
+            pipeline.apply(TextIO.read().from(context.getInputFile()))
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java:45]: This indentation seems incorrect to me.

File: src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java [hunk 19]
```diff
@@ -2,46 +2,60 @@
...
-      .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())))
-      .apply(
+      .apply(ParDo.of(new ConvertLineToVariantFn(vcfParser, context.getHeaderLines())));
+
+    PCollection<TableRow> tableRowPCollection = variantContextPCollection
+            .apply("VariantContextToBQRow",
+                    ParDo.of(new ConvertVariantToRowFn(bigQueryRowGenerator,
+                            context.getVCFHeader())));
+
+    variantContextPCollection.apply(
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java:54]: This was done for Demo purposes, we can remove this part and inline the above row generation into the main pipeline.

For testing purposes, you could create something like the below code to write BqRows into a text file, and verify the results. Right now I think it wouldn't work, at least because I do not see a binding for BqRowGen and BqRowGenImpl added in any of the modules, so maybe some testing would be appropriate.

File: src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGenerator.java [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+import com.google.api.services.bigquery.model.TableRow;
+import htsjdk.variant.variantcontext.VariantContext;
+import htsjdk.variant.vcf.VCFHeader;
+
+/**
+ * Service to generate Big Query Row from VariantContext
+ */
+public interface BigQueryRowGenerator {
+  /**
+   * BigQuery row generator. It provides the common functionalities when generating BigQuery
```
Comment (tneymanov) [inline src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGenerator.java:14]: This comment is for the service, not for the getRows() method. Please modify.

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+      String attrName = entry.getKey();
+      Object value = entry.getValue();
+      VCFInfoHeaderLine infoMetadata = vcfHeader.getInfoHeaderLine(attrName);
+      VCFHeaderLineType infoType = infoMetadata.getType();
+      VCFHeaderLineCount infoCount = infoMetadata.getCountType();
+      if (infoCount == VCFHeaderLineCount.A) {
+        // put this info into ALT field
+        splitAlternateAlleleInfoFields(attrName, value, altMetadata, infoType);
+      } else {
+        row.set(attrName, convertToDefinedType(value, infoType));
```
Comment (samanvp) [inline src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:85]: I am not still sure if we need to implement the logic of converting to type ouerselves. I think it's much safer is we rely on HTSJDK's implementation:
* `getAttributeAsBoolean`
* `getAttributeAsDouble`
* `getAttributeAsDoubleList`
* `getAttributeAsInt`
* `getAttributeAsIntList`
* `getAttributeAsString` 
* `getAttributeAsStringList`

All we need a swtich where calls the corresponding method based on the `infoType` value.

Please let me know what you think.

Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:85]: getAttributeAsInt will return int instead of Integer. We cannot set default value as null. Moreover, those functions only cast the values into target type, like (Integer)object, I do not think it's a safer implementation

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+    // if field value is '.', alternateAllelesList will be empty
+    if (altAlleles.isEmpty()) {
+      TableRow subRow = new TableRow();
+      subRow.set(Constants.ColumnKeyConstants.ALTERNATE_BASES_ALT, null);
+      altMetadata.add(subRow);
+    } else {
+      for (int i = 0; i < altAlleles.size(); ++i) {
+        TableRow subRow = new TableRow();
+        String altBase = altAlleles.get(i).getDisplayString();
+        alleleIndexingMap.put(altBase, i + 1);  // the alt allele index is 1-based
```
[the remark this answers is not in the document]
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:52]: I agree. The ```Genotype``` class does not have this method but ```VariantContext``` does, I'll change the implementation to use this method

File: src/main/java/com/google/gcp_variant_transforms/beam/helper/ConvertLineToVariantFn.java [hunk 23]
```diff
@@ -3,29 +3,32 @@
...
     this.headerLines = headerLines;
   }
 
   @ProcessElement
-  public void processElement(@Element String record, OutputReceiver<Variant> receiver) {
+  public void processElement(@Element String record, OutputReceiver<VariantContext> receiver) {
     // Codec's are not Serializable, so they need to be generated directly on the worker machine.
     if (vcfCodec == null) {
       vcfCodec = vcfParser.generateCodecFromHeaderLines(headerLines);
+      vcfHeader = vcfCodec.getHeader();
```
[the remark this answers is not in the document]
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/beam/helper/ConvertLineToVariantFn.java:30]: Deleted. It serves for previous implementation

File: src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorImpl.java [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.gcp_variant_transforms.common.Constants;
+import htsjdk.variant.variantcontext.GenotypesContext;
+import htsjdk.variant.variantcontext.VariantContext;
+import htsjdk.variant.vcf.VCFHeader;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Base abstract class for BigQuery row generator.
```
[the remark this answers is not in the document]
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/BigQueryRowGeneratorImpl.java:15]: Agree. Changed class comments instead of using the python version comment

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 25]
```diff
@@ -0,0 +1,224 @@
+// Copyright 2020 Google LLC
```
[the remark this answers is not in the document]
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:1]: Got it.

File: src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java (hunk 19 above)
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/beam/VcfToBqPipelineRunner.java:54]: E2E test added in the pipeline runner

File: src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java [hunk 26]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility methods for write values with defined value type into Big Query rows.
+ */
+public class VariantToBqUtils {
+  public static String addReferenceBase(VariantContext variantContext) {
```
[the remark this answers is not in the document]
Comment (BoscoSuen) [reply src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java:26]: To be honest, I am a bit confused why it should be "reference_bases" instead of "reference_base" since it only returns a single string. I will add one "s" in the naming to match to BQ field
